primitives/api/proc-macro/src/decl_runtime_apis.rs

```diff
@@ -323,157 +211,96 @@ fn parse_renamed_attribute(renamed: &Attribute) -> Result<(String, u32)> {
-		result.push(quote!(
-			#[cfg(any(feature = "std", test))]
-			#[allow(clippy::too_many_arguments)]
-			pub fn #fn_name<
-				R: #crate_::Encode + #crate_::Decode + std::cmp::PartialEq,
-				NC: FnOnce() -> std::result::Result<R, #crate_::ApiError> + std::panic::UnwindSafe,
-				Block: #crate_::BlockT,
-				T: #crate_::CallApiAt<Block>,
-			>(
-				call_runtime_at: &T,
-				at: &#crate_::BlockId<Block>,
-				args: std::vec::Vec<u8>,
-				changes: &std::cell::RefCell<#crate_::OverlayedChanges>,
-				storage_transaction_cache: &std::cell::RefCell<
-					#crate_::StorageTransactionCache<Block, T::StateBackend>
-				>,
-				native_call: std::option::Option<NC>,
-				context: #crate_::ExecutionContext,
-				recorder: &std::option::Option<#crate_::ProofRecorder<Block>>,
-			) -> std::result::Result<#crate_::NativeOrEncoded<R>, #crate_::ApiError> {
-				let version = call_runtime_at.runtime_version_at(at)?;
-
-				#(
-					// Check if we need to call the function by an old name.
-					if version.apis.iter().any(|(s, v)| {
-						s == &ID && *v < #versions
-					}) {
-						let params = #crate_::CallApiAtParams::<_, fn() -> _, _> {
-							at,
-							function: #old_names,
-							native_call: None,
-							arguments: args,
-							overlayed_changes: changes,
-							storage_transaction_cache,
-							context,
-							recorder,
-						};
-
-						let ret = #crate_::CallApiAt::<Block>::call_api_at(call_runtime_at, params)?;
-
-						return Ok(ret)
-					}
-				)*
-
-				let params = #crate_::CallApiAtParams {
-					at,
-					function: #trait_fn_name,
-					native_call,
-					arguments: args,
-					overlayed_changes: changes,
-					storage_transaction_cache,
-					context,
-					recorder,
-				};
-
-				#crate_::CallApiAt::<Block>::call_api_at(call_runtime_at, params)
-			}
-		));
-	}
-
-	Ok(quote!( #( #result )* ))
-}
-
 /// Generate the declaration of the trait for the runtime.
 fn generate_runtime_decls(decls: &[ItemTrait]) -> Result<TokenStream> {
 	let mut result = Vec::new();
 
 	for decl in decls {
 		let mut decl = decl.clone();
+		let decl_span = decl.span();
 		extend_generics_with_block(&mut decl.generics);
 		let mod_name = generate_runtime_mod_name_for_trait(&decl.ident);
 		let found_attributes = remove_supported_attributes(&mut decl.attrs);
 		let api_version =
 			get_api_version(&found_attributes).map(|v| generate_runtime_api_version(v as u32))?;
 		let id = generate_runtime_api_id(&decl.ident.to_string());
 
-		let call_api_at_calls = generate_call_api_at_calls(&decl)?;
+		let trait_api_version = get_api_version(&found_attributes)?;
+
+		let mut methods_by_version: BTreeMap<u64, Vec<TraitItemMethod>> = BTreeMap::new();
 
-		// Remove methods that have the `changed_in` attribute as they are not required for the
-		// runtime anymore.
+		// Process the items in the declaration. The filter_map function below does a lot of stuff
+		// because the method attributes are stripped at this point
 		decl.items = decl
 			.items
 			.iter_mut()
```

Comment:
```suggestion
		decl
			.items
			.iter()
```

Reply:
The `iter_mut()` here should stay. The method is modified in the `for_each()`.